leo_filters/src/imu_filter.cpp

```diff
@@ -0,0 +1,267 @@
+
+    if (node["gyro_bias_x"]) {
+      filter_.setAngularVelocityBiasX(node["gyro_bias_x"].as<double>());
+    }
+
+    if (node["gyro_bias_y"]) {
+      filter_.setAngularVelocityBiasY(node["gyro_bias_y"].as<double>());
+    }
+
+    if (node["gyro_bias_z"]) {
+      filter_.setAngularVelocityBiasZ(node["gyro_bias_z"].as<double>());
+    }
+
+  } catch (YAML::BadFile & e) {
+    RCLCPP_ERROR(get_logger(), "IMU bias file doesn't exist.\n");
+    RCLCPP_ERROR(get_logger(), "Creating IMU bias file with current gyrometer bias.\n");
+
+    node["gyro_bias_x"] = filter_.getAngularVelocityBiasX();
+    node["gyro_bias_y"] = filter_.getAngularVelocityBiasY();
+    node["gyro_bias_z"] = filter_.getAngularVelocityBiasZ();
+
+    std::ofstream fout(file_path);
+    fout << node;
+  }
+}
+
+std::string ImuFilter::get_bias_file_path()
+{
+  std::string ros_home;
+  char * ros_home_env;
+  if (ros_home_env = std::getenv("ROS_HOME")) {
+    ros_home = ros_home_env;
+  } else if (ros_home_env = std::getenv("HOME")) {
+    ros_home = ros_home_env;
+    ros_home += "/.ros";
+  }
+
+  return ros_home + "/imu_calibration.yaml";
+}
+
+void ImuFilter::update_filter_params()
+{
+  if (filter_.getDoBiasEstimation() != params_.do_bias_estimation) {
+    filter_.setDoBiasEstimation(params_.do_bias_estimation);
+  }
+
+  if (filter_.getDoAdaptiveGain() != params_.do_adaptive_gain) {
+    filter_.setDoAdaptiveGain(params_.do_adaptive_gain);
+  }
+
+  double gain = filter_.getGainAcc();
+  if (std::fabs(gain - params_.gain_acc) > 0.001) {
+    if (!filter_.setGainAcc(params_.gain_acc)) {
+      RCLCPP_WARN(
+        this->get_logger(),
+        "Invalid gain_acc passed to ComplementaryFilter.");
+    }
+  }
+
+  double bias_alpha = filter_.getBiasAlpha();
+  if (std::fabs(bias_alpha - params_.bias_alpha) > 0.001) {
+    if (!filter_.setBiasAlpha(params_.bias_alpha)) {
+      RCLCPP_WARN(
+        this->get_logger(),
+        "Invalid bias_alpha passed to ComplementaryFilter.");
+    }
+  }
+}
+
+void ImuFilter::check_dynamic_parameters()
+{
+  if (param_listener_.is_old(params_)) {
+    param_listener_.refresh_dynamic_parameters();
+    params_ = param_listener_.get_params();
+    update_filter_params();
+  }
+}
+
+void ImuFilter::imu_callback(sensor_msgs::msg::Imu::SharedPtr msg)
+{
+  const geometry_msgs::msg::Vector3 & a = msg->linear_acceleration;
+  const geometry_msgs::msg::Vector3 & w = msg->angular_velocity;
+  const rclcpp::Time & time = msg->header.stamp;
+
+  check_dynamic_parameters();
+
+  if (!initialized_filter_) {
+    time_prev_ = time;
+    initialized_filter_ = true;
+    return;
+  }
+
+  double dt;
+  if (params_.constant_dt > 0.0) {
+    dt = params_.constant_dt;
+  } else {
+    dt = (time - time_prev_).nanoseconds() * 1e-9;
+  }
+
+  time_prev_ = time;
+
+  filter_.update(a.x, a.y, a.z, w.x, w.y, w.z, dt);
+
+  publish(msg);
+}
+
+tf2::Quaternion ImuFilter::hamiltonToTFQuaternion(
+  double q0, double q1,
+  double q2, double q3) const
```

Comment:
Change this to inline function
```suggestion
inline tf2::Quaternion hamiltonToTFQuaternion(double q0, double q1, double q2, double q3)
```